src/theory/bags/bags_rewriter.cpp

```diff
@@ -0,0 +1,427 @@
+    //        (and (= x y) (= c d)))
+    Node x = n[0][0];
+    Node c = n[0][1];
+    Node y = n[1][0];
+    Node d = n[1][1];
+    NodeManager* nm = NodeManager::currentNM();
+    Node zero = nm->mkConst(Rational(0));
+    Node lte = nm->mkNode(LEQ, c, zero).andNode(nm->mkNode(LEQ, d, zero));
+    Node equal = x.eqNode(y).andNode(c.eqNode(d));
+    Node ite = nm->mkNode(ITE, lte, nm->mkConst(true), equal);
+    return RewriteResponse(REWRITE_AGAIN_FULL, ite);
+  }
+  return RewriteResponse(REWRITE_DONE, n);
+}
+
+RewriteResponse BagsRewriter::rewriteIsIncluded(const TNode& n) const
+{
+  Assert(n.getKind() == BAG_IS_INCLUDED);
+  // (bag.is_included A B) = ((difference_subtract A B) == emptybag)
+  NodeManager* nm = NodeManager::currentNM();
+  Node emptybag = nm->mkConst(EmptyBag(n[0].getType()));
+  Node subtract = nm->mkNode(DIFFERENCE_SUBTRACT, n[0], n[1]);
+  Node equal = subtract.eqNode(emptybag);
+  return RewriteResponse(REWRITE_AGAIN, equal);
+}
+
+RewriteResponse BagsRewriter::rewriteMakeBag(const TNode& n) const
+{
+  Assert(n.getKind() == MK_BAG);
+  // return emptybag for negative or zero multiplicity
+  if (n[1].isConst() && n[1].getConst<Rational>().sgn() != 1)
+  {
+    // (mkBag x c) = emptybag where c <= 0
+    NodeManager* nm = NodeManager::currentNM();
+    Node emptybag = nm->mkConst(EmptyBag(n.getType()));
+    return RewriteResponse(REWRITE_DONE, emptybag);
+  }
+  return RewriteResponse(REWRITE_DONE, n);
+}
+
+RewriteResponse BagsRewriter::rewriteBagCount(const TNode& n) const
+{
+  Assert(n.getKind() == BAG_COUNT);
+  if (n[1].isConst() && n[1].getKind() == EMPTYBAG)
+  {
+    // (bag.count x emptybag) = 0
+    NodeManager* nm = NodeManager::currentNM();
+    return RewriteResponse(REWRITE_DONE, nm->mkConst(Rational(0)));
+  }
+  if (n[1].getKind() == MK_BAG && n[0] == n[1][0])
+  {
+    // (bag.count x (mkBag x c) = c where c > 0 is a constant
+    return RewriteResponse(REWRITE_DONE, n[1][1]);
+  }
+  return RewriteResponse(REWRITE_DONE, n);
+}
+
+RewriteResponse BagsRewriter::rewriteUnionMax(const TNode& n) const
+{
+  Assert(n.getKind() == UNION_MAX);
+  if (n[1].getKind() == EMPTYBAG || n[0] == n[1])
+  {
+    // (union_max A A) = A
+    // (union_max A emptybag) = A
+    return RewriteResponse(REWRITE_DONE, n[0]);
+  }
+  if (n[0].getKind() == EMPTYBAG)
+  {
+    // (union_max emptybag A) = A
+    return RewriteResponse(REWRITE_DONE, n[1]);
+  }
+  if (n[1].getKind() == UNION_MAX && (n[0] == n[1][0] || n[0] == n[1][1]))
+  {
+    // (union_max A (union_max A B) = (union_max A B)
+    // (union_max A (union_max B A) = (union_max B A)
+    return RewriteResponse(REWRITE_DONE, n[1]);
+  }
+  if (n[0].getKind() == UNION_MAX && (n[0][0] == n[1] || n[0][1] == n[1]))
+  {
+    // (union_max (union_max A B) A) = (union_max A B)
+    // (union_max (union_max B A) A) = (union_max B A)
+    return RewriteResponse(REWRITE_DONE, n[0]);
+  }
+
+  if (n[1].getKind() == UNION_DISJOINT && (n[0] == n[1][0] || n[0] == n[1][1]))
```

Comment:
These could perhaps be combined with the cases above, this would make things easier to read.

Reply:
Done.